src/libdredd/src/mutation_replace_binary_operator.cc

```diff
@@ -69,171 +69,203 @@ bool MutationReplaceBinaryOperator::IsValidReplacementOperator(
+                             ->getName(ast_context.getPrintingPolicy())
+                             .str();
+  std::string rhs_type = binary_operator_.getRHS()
+                             ->getType()
+                             ->getAs<clang::BuiltinType>()
+                             ->getName(ast_context.getPrintingPolicy())
+                             .str();
 
-  // Consider every operator apart from the existing operator
-  std::string arg1_evaluated("arg1()");
-  std::string arg2_evaluated("arg2()");
-  for (auto op : operators) {
-    if (op == binary_operator_.getOpcode() || !IsValidReplacementOperator(op)) {
-      continue;
+  if (binary_operator_.isAssignmentOp()) {
+    clang::QualType qualified_lhs_type = binary_operator_.getLHS()->getType();
+    if (qualified_lhs_type.isVolatileQualified()) {
+      assert(binary_operator_.getType().isVolatileQualified() &&
+             "Expected expression to be volatile-qualified since LHS is.");
+      lhs_type = "volatile " + lhs_type;
     }
-    new_function << "  if (__dredd_enabled_mutation(local_mutation_id + "
-                 << mutant_offset << ")) return " << arg1_evaluated << " "
-                 << clang::BinaryOperator::getOpcodeStr(op).str() << " "
-                 << arg2_evaluated << ";\n";
-    mutant_offset++;
   }
-  if (!binary_operator_.isAssignmentOp()) {
-    // LHS
-    new_function << "  if (__dredd_enabled_mutation(local_mutation_id + "
-                 << mutant_offset << ")) return " << arg1_evaluated << ";\n";
-    mutant_offset++;
-
-    // RHS
-    new_function << "  if (__dredd_enabled_mutation(local_mutation_id + "
-                 << mutant_offset << ")) return " << arg2_evaluated << ";\n";
-    mutant_offset++;
-  }
-  if (binary_operator_.isLogicalOp()) {
-    // true
-    new_function << "  if (__dredd_enabled_mutation(local_mutation_id + "
-                 << mutant_offset << ")) return true;\n";
-    mutant_offset++;
-
-    // false
-    new_function << "  if (__dredd_enabled_mutation(local_mutation_id + "
-                 << mutant_offset << ")) return false;\n";
-    mutant_offset++;
-  }
-
-  new_function
-      << "  return " << arg1_evaluated << " "
-      << clang::BinaryOperator::getOpcodeStr(binary_operator_.getOpcode()).str()
-      << " " << arg2_evaluated << ";\n";
-  new_function << "}\n\n";
-
-  // The function captures |mutant_offset| different mutations, so bump up
-  // the mutation id accordingly.
-  mutation_id += mutant_offset;
-
-  return new_function.str();
-}
-
-void MutationReplaceBinaryOperator::Apply(
-    clang::ASTContext& ast_context, const clang::Preprocessor& preprocessor,
-    int first_mutation_id_in_file, int& mutation_id, clang::Rewriter& rewriter,
-    std::unordered_set<std::string>& dredd_declarations) const {
-  std::string new_function_name = "__dredd_replace_binary_operator_";
+  std::string function_name = "__dredd_replace_binary_operator_";
```

Comment:
A good convention is to call the thing you will ultimately return `result`, hence I would rename this to `result`.

Reply:
Renamed.